File: programmers/그래프/순위.cpp
#include <string>
#include <vector>

using namespace std;
int dp[101][101];

int solution(int n, vector<vector<int>> results) {
    int answer = 0;
    for(int i = 0; i < results.size(); i++){
        dp[results[i][0]][results[i][1]] = 1;
    }
    
    // 플로이드와샬 알고리즘(시간복잡도: O(N^3))
    for(int k = 1; k <= n; k++){     // k 는 거쳐가는 정점
        for(int i = 1; i <= n; i++){    // i 는 이긴다.(출발 정점)
            for(int j = 1; j <= n; j++){    // j 는 진다.(도착 정점)
                if(dp[i][k] && dp[k][j])
                    dp[i][j] = 1; 
            }
        }
    }
    
    for(int i = 1; i <= n; i++){
        int sum = 0;
        for(int j = 1; j <= n; j++){
            if(i == j)
                continue;
            
            sum += dp[i][j] + dp[j][i];
        }
        
        if(sum == n-1)  // 한 선수가 모든 선수들에 대해서 (이기든 지든)값을 가진다면 그 선수의 순위를 매길 수 있음
            answer++;
    }
    
    return answer;
}

File: python/플로이드워셜/순위.py
# https://school.programmers.co.kr/learn/courses/30/lessons/49191?language=python3

def solution(n, results):
    answer = 0
    
    # 1. 2차원 리스트 초기화
    dp = [[0] * (n + 1) for _ in range(n + 1)]
    
    # 2. 초기 경기 결과 기록 (a가 b를 이김)
    for a, b in results:
        dp[a][b] = 1
        
    # 3. 플로이드-워셜 알고리즘 핵심 로직
    # k: 거쳐가는 선수, i: 승자, j: 패자
    for k in range(1, n + 1):
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                # i가 k를 이기고, k가 j를 이겼다면 -> i는 j를 이긴다!
                if dp[i][k] == 1 and dp[k][j] == 1:
                    dp[i][j] = 1
                    
    # 4. 정확하게 순위를 매길 수 있는 선수 찾기
    for i in range(1, n + 1):
        count = 0
        for j in range(1, n + 1):
            # 내가 이겼거나(dp[i][j]), 내가 졌거나(dp[j][i]) 승패 관계를 아는 경우
            if dp[i][j] == 1 or dp[j][i] == 1:
                count += 1
                
        # 나를 제외한 모든 선수(n-1명)와의 승패를 안다면 순위 확정 가능
        if count == n - 1:
            answer += 1
            
    return answer